pyocd/target/family/target_nRF52.py

```diff
@@ -0,0 +1,134 @@
+CTRL_AP_IDR = 0x0FC
+
+CTRL_AP_ERASEALLSTATUS_READY = 0x0
+CTRL_AP_ERASEALLSTATUS_BUSY = 0x1
+
+CTRL_AP_APPROTECTSTATUS_ENABLED = 0x0
+CTRL_AP_APPROTECTSTATUS_DISABLED = 0x1
+
+CTRL_AP_RESET_NORESET = 0x0
+CTRL_AP_RESET_RESET = 0x1
+
+CTRL_AP_ERASEALL_NOOPERATION = 0x0
+CTRL_AP_ERASEALL_ERASE = 0x1
+
+CTRL_IDR_EXPECTED = 0x2880000
+CTRL_IDR_VERSION_MASK = 0xf0000000
+CTRL_IDR_VERSION_SHIFT = 28
+
+MASS_ERASE_TIMEOUT = 15.0
+
+LOG = logging.getLogger(__name__)
+
+
+class NRF52(CoreSightTarget):
+
+    VENDOR = "Nordic Semiconductor"
+
+    def __init__(self, link, memoryMap=None):
+        super(NRF52, self).__init__(link, memoryMap)
+        self._svd_location = SVDFile.from_builtin("nrf52.svd")
+        self.ctrl_ap = None
+
+    def create_init_sequence(self):
+        seq = super(NRF52, self).create_init_sequence()
+
+        # Must check whether security is enabled, and potentially auto-unlock, before
+        # any init tasks that require system bus access.
+        seq.insert_before('init_ap_roms',
+                          ('check_ctrl_ap_idr', self.check_ctrl_ap_idr),
+                          ('check_flash_security', self.check_flash_security),
+                          )
+
+        return seq
+
+    def check_ctrl_ap_idr(self):
+        self.ctrl_ap = self.dp.aps[CTRL_AP_NUM]
+
+        # Check CTRL-AP ID.
+        if (self.ctrl_ap.idr & ~CTRL_IDR_VERSION_MASK) != CTRL_IDR_EXPECTED:
+            LOG.error("%s: bad CTRL-AP IDR (is 0x%08x)", self.part_number, self.ctrl_ap.idr)
+
+        ctrl_ap_version = (self.ctrl_ap.idr & CTRL_IDR_VERSION_MASK) >> CTRL_IDR_VERSION_SHIFT
+        LOG.debug("CTRL-AP version %d", ctrl_ap_version)
+
+    def check_flash_security(self):
+        """! @brief Check security and unlock device.
+
+        This init task determines whether the device is locked (APPROTECT enabled). If it is,
+        and if auto unlock is enabled, then perform a mass erase to unlock the device.
+
+        This init task runs *before* cores are created.
+        """
+
+        if self.is_locked():
+            if self.session.options.get('auto_unlock'):
+                LOG.warning("%s APPROTECT enabled: will try to unlock via mass erase", self.part_number)
+
+                # Do the mass erase.
+                if not self.mass_erase():
+                    LOG.error("%s: mass erase failed", self.part_number)
+                    raise exceptions.TargetError("unable to unlock device")
+                # Cached badness from create_ap run during AP lockout prevents create_cores from
+                # succeeding.
+                self.dp.create_1_ap(AXI_AP_NUM)
+            else:
+                LOG.warning("%s APPROTECT enabled: not automatically unlocking", self.part_number)
+        else:
+            LOG.info("%s not in secure state", self.part_number)
+
+    def is_locked(self):
+        if not self.ctrl_ap:
+            if CTRL_AP_NUM not in self.aps:
+                self.dp.create_1_ap(CTRL_AP_NUM)
```

Comment:
Is this really necessary? It doesn't hurt, so I'm mostly curious. And if you need it here, wouldn't it also be needed on `mass_erase()`?

Reply:
It's not necessary, but without it the *status* command in commander throws an exception in no_init mode.

I didn't have it initially, but it seems like a lot to ask of a (potentially new) user to know the chip specifics (that another AP exists) and run `makeap 1` before *status* works.

Now that you mention it, if the user runs *unlock* before *status*, it'll fail in the same way.

They should both go the same way, but I'll defer you you on which way that is.

Reply:
I promoted this logic to a separate method *_ensure_ctrl_ap*, and referenced it in both *mass_erase* and *is_locked*. The more I think about it, the more I think that commander should just "do the right thing".

I cannot think of a reason why the user would give us the specific target, but then *not* want to init the AP when needed for a requested operation.

Reply:
`--no-init` is primarily meant for bring-up of new chips and debugging low-level issues. So having commander just do the right thing, at least in this case, may be problematic as it may cause unexpected side effects. The way I've been using `--no-init` (and sort of how I meant it to be used) is to manually create an AP before any operation that requires it by running the `makeap` command. (This is similar to how you need to run the `initdp` command first in no-init mode. Of course, I realise none of this is documented… ☹️)

Still, I'm happy to merge the PR this way if that's what you prefer.

Reply:
Let's do it your way, but can you wait on the merge? 

I'll look into adding some documentation around using commander for manual unlock.

Kinetis is similar enough that such documentation would serve both.

Reply:
Yes, of course, I'm happy to wait til you're ready!